hedera-node/src/main/java/com/hedera/services/ledger/interceptors/EndOfStakingPeriodCalculator.java

```diff
@@ -0,0 +1,149 @@
+import com.hedera.services.context.properties.PropertySource;
+import com.hedera.services.records.RecordsHistorian;
+import com.hedera.services.state.EntityCreator;
+import com.hedera.services.state.merkle.MerkleAccount;
+import com.hedera.services.state.merkle.MerkleNetworkContext;
+import com.hedera.services.state.merkle.MerkleStakingInfo;
+import com.hedera.services.store.contracts.precompile.SyntheticTxnFactory;
+import com.hedera.services.utils.EntityNum;
+import com.hederahashgraph.api.proto.java.NodeStake;
+import com.hederahashgraph.api.proto.java.Timestamp;
+import com.swirlds.merkle.map.MerkleMap;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static com.hedera.services.records.TxnAwareRecordsHistorian.DEFAULT_SOURCE_ID;
+import static com.hedera.services.state.EntityCreator.NO_CUSTOM_FEES;
+
+@Singleton
+public class EndOfStakingPeriodCalculator {
+	public static final String END_OF_STAKING_PERIOD_CALCULATIONS_MEMO = "End of Staking Period Calculation record";
+
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfoSupplier;
+	private final Supplier<MerkleNetworkContext> merkleNetworkContextSupplier;
+	private final SyntheticTxnFactory syntheticTxnFactory;
+	private final RecordsHistorian recordsHistorian;
+	private final EntityCreator creator;
+	private final PropertySource properties;
+
+	@Inject
+	public EndOfStakingPeriodCalculator(
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
+			final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfoSupplier,
+			final Supplier<MerkleNetworkContext> merkleNetworkContextSupplier,
+			final SyntheticTxnFactory syntheticTxnFactory,
+			final RecordsHistorian recordsHistorian,
+			final EntityCreator creator,
+			@CompositeProps PropertySource properties
+	) {
+		this.accounts = accounts;
+		this.stakingInfoSupplier = stakingInfoSupplier;
+		this.merkleNetworkContextSupplier = merkleNetworkContextSupplier;
+		this.syntheticTxnFactory = syntheticTxnFactory;
+		this.recordsHistorian = recordsHistorian;
+		this.creator = creator;
+		this.properties = properties;
+	}
+
+	public void updateNodes(final Instant consensusTime) {
+		final var stakingInfo = stakingInfoSupplier.get();
+		final var merkleNetworkContext = merkleNetworkContextSupplier.get();
+
+		// skip end of staking period calculations if the rewards are not yet activated.
+		if (!merkleNetworkContext.areRewardsActivated()) {
+			return;
+		}
+
+		final var rewardRate = Math.min(stakingRewardsAccountBalance(),
+				properties.getLongProperty("staking.rewardRate"));
+
+		long updatedTotalStakedRewardStart = 0L;
+		long updatedTotalStakedStart = 0L;
+		List<NodeStake> nodeStakingInfos = new ArrayList<>();
+
+		for (final var nodeNum : stakingInfo.keySet()) {
```

Comment:
Since this only happens once a day, performance doesn't matter, so to improve readability I might add a `stakingInfo.keySet().stream().sorted().toList()` here.